### PR TITLE
SI-6135 SI-6443 Bridges for dependent method types

### PR DESCRIPTION
**Pull request by retronym**
Bridge building operates on unusual method signatures:
after uncurry, so parameter lists are collapsed; but before
erasure, so dependently typed parameters are still around.

Original:

```
def foo(a: T)(b: a.type, c: a.U): Unit
```

During computeBridges:

```
(a: T, b: a.type, c: a.U)Unit
```

This signature no longer appears to override the corresponding
one in a superclass, because the types of `b` and `c` are dependent
on method parameters.

This commit instead checks for overriding pairs with fully
curried signatures, such as:

```
(a: T)(b: a.type)(c: a.U)Unit
```

Existing logic in [`matchesQuantified`](https://github.com/retronym/scala/blob/8e8aa4b1d/src/reflect/scala/reflect/internal/Types.scala#L6175), a local method in
`matchesType`, then correctly substitutes the superclass signature
parameter symbol into the subclass signature.

Review by @odersky

### CODE REVIEW

File: src/compiler/scala/tools/nsc/transform/OverridingPairs.scala [hunk 1]
```diff
@@ -46,14 +46,17 @@ abstract class OverridingPairs {
...
+      def tp_s(s: Symbol) = {
+        val memType = memberType(self, s)
+        s"${memType}/${memType.getClass}"
+      }
+      val result = sym1.isType || (memberType(self, sym1) matches memberType(self, sym2))
+      debuglog(s"overriding-pairs? ${sym1.fullLocationString} matches ${sym2.fullLocationString} (${tp_s(sym1)} vs. ${tp_s(sym2)}) == $result")
       result
     }
 
+    protected def memberType(self: Type, sym: Symbol) = self memberType sym
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/transform/OverridingPairs.scala]: Maybe a different name would be more appropriate to emphasize the difference between `Type.memberType` and the thing here?

Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/OverridingPairs.scala]: done.

**Comment by retronym** [comment]
PLS REBUILD ALL

**Comment by retronym** [comment]
> Review by @odersky

Or @adriaanm

**Comment by adriaanm** [comment]
will perform my neglected reviewing duties this afternoon

**Comment by retronym** [comment]
I discussed this issue problem briefly with Martin today. He proposed an alternative: if `uncurry` wants to concatenate the parameter lists it should take measures to avoid leaving signatures like `def foo(a: Int, b: a.type)` which cannot be expressed as Scala source. Presumably this would require erasing the singleton types to their bounds (do we have a type map for that?) and inserting casts in the body of the method.

It's a bigger change than this one, but it does seem more principled. @adriaanm: does it sound viable?

**Comment by retronym** [comment]
Another question that springs to mind: why uncurry in `uncurry`, as opposed to doing it later on (say, in `erasure`)?

**Comment by adriaanm** [comment]
maybe we could do the override checking before uncurry?

could we change `beforeExplicitOuter` to `beforeUncurry` in:

```
    val opc = beforeExplicitOuter {
      new overridingPairs.Cursor(root) {
        override def parents              = List(root.info.firstParent)
        override def exclude(sym: Symbol) = !sym.isMethod || sym.isPrivate || super.exclude(sym)
      }
    }
```

**Comment by retronym** [comment]
Key moments in the evolution of that hunk:

https://github.com/scala/scala/commit/3761cb4b#L9L370

``` patch
@@ -367,7 +385,19 @@ abstract class Erasure extends AddInterfaces with typechecker.Analyzer {
       val bridgesScope = new Scope();
       val bridgeTarget = new HashMap[Symbol, Symbol];
       var bridges: List[Tree] = List();
-      for (val bc <- site.baseClasses.tail; val other <- bc.info.members) {
+      val opc = atPhase(phase.prev) {  // to avoid DEFERRED flags for interfaces
+       new overridingPairs.Cursor(owner) {
+          override def parents: List[Type] = List(owner.info.parents.head);
+          override def exclude(sym: Symbol): boolean =
+            !sym.isMethod || (sym hasFlag (PRIVATE | BRIDGE)) || super.exclude(sym);
+        }
+      }
+      while (opc.hasNext) {
+       val member = opc.overriding;
+       val other = opc.overridden;
+       if (!(member hasFlag DEFERRED)) {
```

https://github.com/scala/scala/commit/cda2954#L3L488

``` patch
-      val opc = atPhase(phase.prev) {         // to avoid DEFERRED flags for interfaces
-        new overridingPairs.Cursor(owner) {
+      val opc = atPhase(phase.prev.prev) {     // bq: who understands prev comment "to avoid DEFERRED flags for interfaces"
+        new overridingPairs.Cursor(owner) {    //     it seems we want types *before* explicitOuter
```

https://github.com/scala/scala/commit/def54abfb#L2L488

``` patch
diff --git a/src/compiler/scala/tools/nsc/transform/Erasure.scala b/src/compiler/scala/tools/nsc/transform/Erasure.scala
index 4aa3d2f..ef1f066 100644
--- a/src/compiler/scala/tools/nsc/transform/Erasure.scala
+++ b/src/compiler/scala/tools/nsc/transform/Erasure.scala
@@ -485,8 +485,8 @@ abstract class Erasure extends AddInterfaces with typechecker.Analyzer {
       val bridgesScope = newScope
       val bridgeTarget = new HashMap[Symbol, Symbol]
       var bridges: List[Tree] = List()
-      val opc = atPhase(phase.prev.prev) {     // bq: who understands prev comment "to avoid DEFERRED flags for interfaces"
-        new overridingPairs.Cursor(owner) {    //     it seems we want types *before* explicitOuter
+      val opc = atPhase(currentRun.explicitOuterPhase) {
+        new overridingPairs.Cursor(owner) {
           override def parents: List[Type] = List(owner.info.parents.head)
```

I also wonder: if we go back `beforeUncurry`, will we neglect to install bridge in artifacts of specialization?

**Comment by adriaanm** [comment]
excellent point: members synthesized by specialization wouldn't be part of the class's decls before uncurry,
but you could try  only overriding `protected def matches(sym1: Symbol, sym2: Symbol): Boolean` in `Cursor` and wrap its super-call in a beforeUncurry

**Comment by retronym** [comment]
But wouldn't `beforeUncurry` be prehistoric for said synthesized members?

**Comment by paulp** [comment]
Here's a thought: specialization is unfixable, and we will jettison it eventually anyway. We can let it burn off a few thousand of our hours (to go with the at least few thousand it has already collected) and THEN jettison it, or we could keep those few thousand hours for our own use.

Every time you try to reason your way to how something will interact with specialization, god kills a kitten.

**Comment by adriaanm** [comment]
@retronym, it depends on how those members are synthesized -- i haven't checked whether they are created atPhase(typer) or something later (which would imply those prehistoric woes), but as @paulp says: don't kill kittens (OTOH, actively killing specialization isn't great either -- it does seem we've run out of steam walking on those specialized egg shells)

**Comment by retronym** [comment]
Righto, I'll see what I can find. But can we agree that uncurry isn't playing by the rules with regards to generating trees that don't correspond to legal Scala?

The rewriting I had in mind was:

``` scala
def foo(a: A, b: B)(c: a.M[b.type]) = b

// after uncurry
def foo(a: A, b: B, c: (_a.M[_b.type]) forSome { val _a: A; val _b: B }) = { <synth> val c$1 = c.asInstanceOf[a.M[b.type]; c$1 }
```

**Comment by adriaanm** [comment]
``` scala
scala> trait A {trait M[x]} ; trait B
defined trait A
defined trait B

scala> class X { def foo(a: A, b: B)(c: a.M[b.type]) = b }
defined class X

scala> val methTp = intp("X").info.decls.toList(1).info
methTp: $r.intp.global.Type = (a: A, b: B)(c: a.M[b.type])B

scala> existentialAbstraction(methTp.paramss.flatten, methTp)
res2: $r.intp.global.Type = ((a: A, b: B)(c: a.M[b.type])B) forSome { a: A; b: B }

```

**Comment by adriaanm** [comment]
it's not entirely sanitary (as in respecting scoping of method param symbols), but should illustrate one potential solution

**Comment by retronym** [comment]
I'm going to close this and try out the alternative approaches.

**Comment by retronym** [comment]
The code is still a complete mess, but this approach seems to work:

https://github.com/retronym/scala/compare/ticket/6443-uncurry-surgery

``` scala
          def foo(a: A, b: B, c: A#M[B]): B = {
            <synthetic> val c$1: a.M[b.type] = c.asInstanceOf[a.M[b.type]]();
            b
          }
```

```
ticket/6443-uncurry-surgery ~/code/scala ./test/partest test/files/run/t6443*.scala test/files/neg/t6443*.scala test/files/run/t6135.scala 

Testing individual files
testing: [...]/files/neg/t6443c.scala                                 [  OK  ]

Testing individual files
testing: [...]/files/run/t6135.scala                                  [  OK  ]
testing: [...]/files/run/t6443b.scala                                 [  OK  ]
testing: [...]/files/run/t6443.scala                                  [  OK  ]
All of 4 tests were successful (elapsed time: 00:00:10)
```

**Pull request closed without merging**